### ISSUE TITLE
CUBLAS_STATUS_NOT_INITIALIZED

### ISSUE THREAD

**Issue by bvandenbon** [lightvector/KataGo#22]
installed katago, works fine most of the time. But sometimes I have the following error.
But as you can see, just retrying fixes it.

![afbeelding](https://user-images.githubusercontent.com/20482760/59804067-a1be0480-92ed-11e9-9e9b-5803d209a513.png)

**Comment by bvandenbon** [comment]
Actually, now also getting this one : 

![afbeelding](https://user-images.githubusercontent.com/20482760/59806782-e64d9e00-92f5-11e9-9fcf-d16f30bac238.png)

(which isn't solved by a simple retry).

**Comment by lightvector** [comment]
Both errors are nondeterministic but sometimes it works all the way to the point of successfully playing a move? If so, that makes things a bit hard to debug, particularly if I don't know how to reproduce it. Is there any more information you can give or are you basically stuck at this point?

**Comment by bvandenbon** [comment]
When I first tried katago, it worked straightaway until the point of playing and analyzing. It worked flawless. 

Then a couple of hours later, I tried again, and had the CUBLAS_STATUS_NOT_INITIALIZED error. I just retried and it worked. 

Some minutes later had it again, retried and it worked.

Then after some hours I wanted to test some more, and I had the CUDNN_STATUS_INTERNAL_ERROR issue. Retrying didn't help this time.

But then (the next day) just a couple of minutes ago, (without reboots or anything special) I tried again and everything worked.

Having said all that. There are other services running on this server. It runs multiple LeeLa Zero instances, and sometimes starts/stops instances on other GPUs. And so, in a way, the conditions change constantly. For LeeLa Zero there wasn't a problem when multiple applications used the same GPU. Apparently that worked fine.

But maybe (just guessing here), maybe the CUDA integration of KataGo is maybe more restrictive in regards to sharing of resources.  Next time when I have this issue, I will stop all other services, and will test if this is relevant. 

A related question that comes to mind: Is there a way to specify the GPU as we can with LeeLa Zero? LeeLa Zero has a switch to set the GPU (e.g. booting leelaz with -g6 switch will enable GPU 6.)

**Comment by lightvector** [comment]
Okay, let me know if there's anything I can do!
Right now, the way you specify the set of GPUs you want to use is in the config, such as gtp_example.cfg.
Take a look at the "CUDA GPU settings" section. For example, if you did "cudaGpuToUse = 3" you would use GPU number 3.

To give a more complex example, if you set numNNServerThreadsPerModel = 2 in the section just above it, and then you set both of "cudaGpuToUseThread0=2" and "cudaGpuToUseThread1=5" that would run KataGo using GPU number 2 and GPU number 5 simultaneously, splitting the work between the two. To effectively make use of both GPUs, you would also want to make sure the search was run with multiple CPU threads, by editing "numSearchThreads = 1" to a higher value like 3 or 4.

So basically like this you can configure KataGo to use any number of GPUs or any desired subset of them.

**Comment by bvandenbon** [comment]
Great stuff :)

**Comment by lightvector** [comment]
Not sure the exact cause of the CUDA errors was ever pinpointed, but it seems like a load-related thing and not obviously KataGo-specific. Everything else here seems handled? So closing this issue. Feel free to reply back if there's anything more! :)
